modules/tinymce/src/plugins/lists/main/ts/core/Util.ts

```diff
@@ -1,25 +1,55 @@
 import { Arr } from '@ephox/katamari';
 
 import Editor from 'tinymce/core/api/Editor';
+import { NodeChangeEvent } from 'tinymce/core/api/EventTypes';
 
 import * as NodeType from './NodeType';
 
 export const isCustomList = (list: HTMLElement): boolean =>
   /\btox\-/.test(list.className);
 
-export const listState = (editor: Editor, listName: string, activate: (active: boolean) => void): () => void => {
-  const nodeChangeHandler = (e: { parents: Node[] }) => {
-    const inList = Arr.findUntil(e.parents, NodeType.isListNode, NodeType.isTableCellNode)
-      .filter((list: HTMLElement) => list.nodeName === listName && !isCustomList(list))
-      .isSome();
-    activate(inList);
+export const setupToggleButtonHandler = (editor: Editor, listName: string) => (api): () => void => {
+  const toggleButtonHandler = (e: NodeChangeEvent) => {
+    api.setActive(inList(e.parents, listName));
+    api.setEnabled(isEditableSelection(editor, e.element));
   };
+  return setNodeChangeHandler(editor, toggleButtonHandler, getInitial(editor));
+};
 
-  // Set the initial state
-  const parents = editor.dom.getParents(editor.selection.getNode());
-  nodeChangeHandler({ parents });
+export const setupMenuButtonHandler = (editor: Editor, listName: string) => (api): () => void => {
+  const menuButtonHandler = (e: NodeChangeEvent) =>
+    api.setEnabled(inList(e.parents, listName) && isEditableSelection(editor, e.element));
+  return setNodeChangeHandler(editor, menuButtonHandler, getInitial(editor));
+};
 
-  editor.on('NodeChange', nodeChangeHandler);
+const getInitial = (editor: Editor): NodeChangeEvent => {
+  const initialNode = editor.selection.getNode();
+  return {
+    parents: editor.dom.getParents(initialNode),
+    element: initialNode
+  };
+};
 
+const setNodeChangeHandler = (editor: Editor, nodeChangeHandler: (e: NodeChangeEvent) => void, initial: NodeChangeEvent): () => void => {
```

Comment:
Since the way to get `initial` is the same for both, why not just call it inside this function?

Our style tends to be defining functions before they are used, so I'd appreciate if you could reorder your changes in this module

Reply:
Will do.